crates/sui/src/client_commands.rs

```diff
@@ -840,6 +854,66 @@ impl SuiClientCommands {
                 };
                 SuiClientCommandResult::Addresses(output)
             }
+            SuiClientCommands::Balance {
+                address,
+                coin_type,
+                with_coins,
+            } => {
+                let address = get_identity_address(address, context)?;
+                let client = context.get_client().await?;
+                let mut objects: Vec<Coin> = Vec::new();
+                let mut cursor = None;
+
+                loop {
+                    let response = client
+                        .coin_read_api()
+                        .get_all_coins(address, cursor, None)
+                        .await?;
+                    objects.extend(response.data);
+
+                    if response.has_next_page {
+                        cursor = response.next_cursor;
+                    } else {
+                        break;
+                    }
+                }
+
+                let mut coins_by_type: BTreeMap<String, (Option<SuiCoinMetadata>, Vec<Coin>)> =
+                    BTreeMap::new();
+                for c in objects {
+                    match coins_by_type.entry(c.coin_type.clone()) {
+                        Entry::Occupied(mut e) => {
+                            e.get_mut().1.push(c);
+                        }
+                        Entry::Vacant(e) => {
+                            let coin_metadata = client
+                                .coin_read_api()
+                                .get_coin_metadata(c.coin_type.clone())
+                                .await
+                                .with_context(|| {
+                                    format!(
+                                        "Cannot fetch the coin metadata for coin {}",
+                                        c.coin_type
+                                    )
+                                })?;
+
+                            let _ = &e.insert((coin_metadata, vec![c]));
+                        }
+                    };
+                }
+
+                let coins = if let Some(coin_type) = coin_type {
```

Comment:
This seems the wrong way round to me -- you are getting all the coins every time, and then dropping the ones you don't want on the floor, instead of only requesting the coins you want in the first place. How about using `get_coins` instead of `get_all_coins` if a `coin_type` is supplied?

Reply:
Fixed!